packages/experiments/src/implementation.js

```diff
@@ -51,15 +60,15 @@ export const __dangerousOptInToUnstableAPIsOnlyForCoreModules = (
 				'your product will inevitably break on one of the next WordPress releases.'
 		);
 	}
-	if ( registeredExperiments.includes( moduleName ) ) {
-		throw new Error(
-			`You tried to opt-in to unstable APIs as module "${ moduleName }" which is already registered. ` +
-				'This feature is only for JavaScript modules shipped with WordPress core. ' +
-				'Please do not use it in plugins and themes as the unstable APIs will be removed ' +
-				'without a warning. If you ignore this error and depend on unstable features, ' +
-				'your product will inevitably break on one of the next WordPress releases.'
-		);
-	}
+	// if ( registeredExperiments.includes( moduleName ) ) {
+	// 	throw new Error(
+	// 		`You tried to opt-in to unstable APIs as module "${ moduleName }" which is already registered. ` +
+	// 			'This feature is only for JavaScript modules shipped with WordPress core. ' +
+	// 			'Please do not use it in plugins and themes as the unstable APIs will be removed ' +
+	// 			'without a warning. If you ignore this error and depend on unstable features, ' +
+	// 			'your product will inevitably break on one of the next WordPress releases.'
+	// 	);
+	// }
```

Comment:
This check doesn't play well with Story Book. There, `__dangerousOptInToUnstableAPIsOnlyForCoreModules` is called multiple times assumedly due to hot module loading. This causes an error when navigating to a different component.

![image](https://user-images.githubusercontent.com/612155/213085518-f94b551d-6772-4fe3-9aee-03d0df03b78a.png)

I'm not sure what to do to get around this. Thoughts @adamziel?

Reply:
Could this be unrelated to HMR? I wonder if the storybook is loading the same file twice.

If this is indeed a hot reloading issue:

Ideally `__dangerousOptInToUnstableAPIsOnlyForCoreModules` wouldn't be called for the second time on hot reload. The module has already opted-in and the global state knows it, there's no need to go through the motions all over again. The easiest solution would be:

```js
try {
    __dangerousOptInToUnstableAPIsOnlyForCoreModules()
} catch(e) {
    // .. ignore
}
```
However, this would obscure genuine errors.

I have two ideas:

1. Do not re-evaluate experiments.js during HMR
2. Detect we're inside HMR and bale out

Reply:
Will defer to you on this 😅 I'm not really sure what's going on. I'll update the PR with your `! IS_GUTENBERG_PLUGIN` suggestion.

Reply:
Oh, that doesn't fix this because `IS_GUTENBERG_PLUGIN` is not set when running the story book (`npm run storybook:dev`).

@ciampo: Is it safe to set `process.env.IS_GUTENBERG_PLUGIN = 1` in `storybook/main.js`?

Reply:
> Could this be unrelated to HMR? I wonder if the storybook is loading the same file twice.

Yes you're right it does seem to be due to story book lazily loading script dependencies. Just a guess: by calling into `__dangerousOptInToUnstableAPIsOnlyForCoreModules` which mutates a global object are we now violating the `"sideEffects": true` condition that e.g. `@wordpress/components` has?

Reply:
I'm not sure about the proper solution right now, but if we want an easy way out, the suggested `try catch` could just ignore the specific errors and not the rest..

Reply:
> I'm not sure about the proper solution right now, but if we want an easy way out, the suggested `try catch` could just ignore the specific errors and not the rest..

Oh I meant if someone tampers with the opt-in process in their plugin, we want to throw an exception. However, in the storybook context, we don't. But it's the same exception.

Reply:
> sideEffects": true

@noisysocks a global state is definitely a side effect. Fortunately, there's no good reason to ever re-evaluate `experiments.js` – if we can make the tooling only run it once, we're good. One idea is to set `"sideEffects": true` in the experiments package and `"sideEffects": ["./src/experiments.js"]` in any package opting-in to experiments ([webpack docs say files arrays are valid values](https://webpack.js.org/guides/tree-shaking/)).

Reply:
I configured `sideEffects` with your suggestions but that didn't seem to affect anything.

I can't figure this out 😢 @adamziel: can you help? You can see the error by:

1. `npm run storybook:dev`
2. Use left sidebar to navigate to the _Block Editor_ story.
3. Use left sidebar to navigate to the _CustomSelectControl_ story.

Reply:
> I configured `sideEffects` with your suggestions but that didn't seem to affect anything.

I can't get it to work either! Let's just remove that check from the Gutenberg plugin with `IS_GUTENBERG_PLUGIN` or a custom constant like `IS_WORDPRESS_CORE` in case we can't use the former in the storybook.

> I also tried wrapping the call to `__dangerousOptInToUnstableAPIsOnlyForCoreModules` in a gate (`if ( ! hasLoaded ) { }`) to no effect.

That would likely have to be `! window.hasLoaded` – I assume all the local variables get wiped out.

Reply:
>  Is it safe to set process.env.IS_GUTENBERG_PLUGIN = true in storybook/main.js?

I _think_ so, although I'd also like to double-check with @mirka and @gziolo (when he's back) to make sure!

Reply:
OK, I've added a check for `IS_GUTENBERG_PLUGIN` and defined this var in `storybook/main.js`.

Reply:
> Is it safe to set process.env.IS_GUTENBERG_PLUGIN = true in storybook/main.js?

Just so I understand this, does the `IS_GUTENBERG_PLUGIN` env var really have no other effect on the behavior of wp-components (and wp-block-editor I guess), and can we be sure that it will stay that way in the future?

Basically this goes against the isolation work we've been doing to untangle implicit Gutenberg/WP dependencies from the default Storybook environment. It also means that setting `IS_GUTENBERG_PLUGIN=true` is now a requirement for getting these packages to work, at least in some environments that are HMR'ed or chunked or whatever the root cause is.

Even if it's "safe", it doesn't make sense from an architectural standpoint, so I think it's something worth resolving more cleanly.

Reply:
@noisysocks in this case I'd go for `ALLOW_EXPERIMENTS_REREGISTRATION` that would be `true` everywhere except WordPress core merge

Reply:
It feels overkill to me that there should be a global env variable for this one specific thing. Can we kill the reregistration check altogether? In my mind the only check that matters is the consent string.

Reply:
> It feels overkill to me that there should be a global env variable for this one specific thing. 

To me it's the opposite – removing that check just for this one specific thing feels like overkill.

The consent string is a carrot and this checks is a stick, I think it's important to have both. A global variable is just a feature flag – they are commonly used in similar scenarios. Why would it be overkill? I could prepare the necessary commit if you wish.

Reply:
@adamziel I'm coming late to this discussion so I don't have all the context -- but it seems to me that allowing double registration of the same module is just fine. We check if the `moduleName` is one of the core modules, and if it is, return the `{ lock, unlock }` pair. We can do the same on the second or the 100th call, there's no reason to allow just a single call.

The `registeredExperiments` array could be removed completely. It's used only to check for these repeated registrations and nothing else. And we could get rid of `resetRegisteredExperiments`, too, making the experiments module stateless.

Reply:
> The consent string is a carrot and this checks is a stick, I think it's important to have both. A global variable is just a feature flag – they are commonly used in similar scenarios. Why would it be overkill? I could prepare the necessary commit if you wish.

Hmmm ok I suppose. I just dislike global config 😅

> The `registeredExperiments` array could be removed completely. It's used only to check for these repeated registrations and nothing else. And we could get rid of `resetRegisteredExperiments`, too, making the experiments module stateless.

Making experiments stateless is appealing to me. I imagine we'll face this exact problem again in a different context that requires loading modules multiple times e.g. adding HMR, adding lazy loading, turning WP Admin into more of a SPA.

Up to you @adamziel, it's your baby 😀

Reply:
> @adamziel I'm coming late to this discussion so I don't have all the context -- but it seems to me that allowing double registration of the same module is just fine.

@jsnajdr The double registration protection is to prevent developers from opting-in to experiments by pretending to be one of the core packages. Everything would still work without it, but the private APIs would be less protected.

> I imagine we'll face this exact problem again (...) adding HMR, adding lazy loading, turning WP Admin into more of a SPA.

IMHO Including strict checks in the initial launch will discourage tampering more effectively than not including them.  Of course – the time may come when they have to be removed. That's fine. It's not like we'll advertise the adjustment anywhere. I'll update this PR to include the additional feature flag.